libs/zzz/formula/src/util.ts

```diff
@@ -0,0 +1,182 @@
+import { cmpEq, cmpNE } from '@genshin-optimizer/pando/engine'
+import type {
+  CharacterKey,
+  DiscMainStatKey,
+  DiscSetKey,
+  DiscSubStatKey,
+  ModificationKey,
+  PhaseKey,
+  WengineKey,
+} from '@genshin-optimizer/zzz/consts'
+import {} from '@genshin-optimizer/zzz/db'
+import type { Member, Preset, TagMapNodeEntries } from './data/util'
+import {
+  convert,
+  getStatFromStatKey,
+  own,
+  ownBuff,
+  ownTag,
+  reader,
+} from './data/util'
+
+export function withPreset(
+  preset: Preset,
+  ...data: TagMapNodeEntries
+): TagMapNodeEntries {
+  return data.map(({ tag, value }) => ({ tag: { ...tag, preset }, value }))
+}
+export function withMember(
+  src: Member,
+  ...data: TagMapNodeEntries
+): TagMapNodeEntries {
+  return data.map(({ tag, value }) => ({ tag: { ...tag, src }, value }))
+}
+
+type TempICharacter = {
+  key: CharacterKey
+  level: number
+  promotion: number
+  basic: number
+  dodge: number
+  special: number
+  chain: number
+  assist: number
+  core: number
+  mindscape: number
+}
+export function charTagMapNodeEntries(data: TempICharacter): TagMapNodeEntries {
+  const {
+    lvl,
+    promotion,
+    basic,
+    dodge,
+    special,
+    chain,
+    assist,
+    core,
+    mindscape,
+  } = own.char
+  const { char, iso, [data.key]: sheet } = reader.withAll('sheet', [])
+
+  return [
+    char.reread(sheet),
+    iso.reread(sheet),
+
+    lvl.add(data.level),
+    basic.add(data.basic),
+    dodge.add(data.dodge),
+    special.add(data.special),
+    chain.add(data.chain),
+    assist.add(data.assist),
+    core.add(data.core),
+    promotion.add(data.promotion),
+    mindscape.add(data.mindscape),
+
+    // Default char
+    ownBuff.base.crit_.add(0.05),
+    ownBuff.base.crit_dmg_.add(0.5),
+  ]
+}
+
+export function wengineTagMapNodeEntries(
+  key: WengineKey,
+  level: number,
+  modification: ModificationKey,
+  phase: PhaseKey
+): TagMapNodeEntries {
+  return [
+    // Mark wengine cones as used
+    own.common.count.sheet(key).add(1),
+    own.wengine.lvl.add(level),
+    own.wengine.modification.add(modification),
+    own.wengine.phase.add(phase),
+  ]
+}
+
+export function discTagMapNodeEntries(
+  stats: Partial<Record<DiscMainStatKey | DiscSubStatKey, number>>,
+  sets: Partial<Record<DiscSetKey, number>>
+): TagMapNodeEntries {
+  const {
+    common: { count },
+    initial,
+  } = convert(ownTag, { sheet: 'disc', et: 'own' })
+  return [
+    // Opt-in for artifact buffs, instead of enabling it by default to reduce `read` traffic
+    reader.sheet('agg').reread(reader.sheet('disc')),
+
+    // Add `sheet:dyn` between the stat and the buff so that we can `detach` them easily
+    // TODO: 4p buffs will go into 'combat' instead of 'initial'
```

Comment:
Noteable todo, might have to remove this reread? @lantua

Reply:
If the solver varies `{ sheet: 'disc', qt: 'initial' }` stats, we should keep it. Maybe add more `dyn` rereads even if the solver also varies other tags.